File: src/core/Shared/AggregateEvent.cs
using System;

namespace core.Shared
{
    public class AggregateEvent
    {
        public AggregateEvent(Guid id, Guid aggregateId, DateTimeOffset when)
        {
            if (id == Guid.Empty)
            {
                throw new InvalidOperationException("id cannot be empty");
            }

            if (aggregateId == Guid.Empty)
            {
                throw new InvalidOperationException("aggregateId cannot be empty");
            }

            Id = id;
            AggregateId = aggregateId;
            When = when;
        }

        public Guid Id { get; }
        public Guid AggregateId { get; }
        public DateTimeOffset When { get; }
    }
}

File: src/core2.fs/Shared/AggregateEvent.fs
namespace core.Shared

open System

[<AllowNullLiteral>]
type AggregateEvent(id: Guid, aggregateId: Guid, ``when``: DateTimeOffset) =
    do
        if id = Guid.Empty then
            invalidOp "id cannot be empty"
        if aggregateId = Guid.Empty then
            invalidOp "aggregateId cannot be empty"
    
    member val Id = id with get
    member val AggregateId = aggregateId with get
    member val When = ``when`` with get
